client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -372,168 +503,76 @@ private static URI buildUri(String path, Map<String, String> params) {
-         */
-        private Builder(HttpHost... hosts) {
-            if (hosts == null || hosts.length == 0) {
-                throw new IllegalArgumentException("no hosts provided");
-            }
-            this.hosts = hosts;
+        FailureTrackingResponseListener(ResponseListener responseListener) {
+            this.responseListener = responseListener;
         }
 
-        /**
-         * Sets the maximum timeout (in milliseconds) to honour in case of multiple retries of the same request.
-         * {@link #DEFAULT_MAX_RETRY_TIMEOUT_MILLIS} if not specified.
-         *
-         * @throws IllegalArgumentException if maxRetryTimeoutMillis is not greater than 0
-         */
-        public Builder setMaxRetryTimeoutMillis(int maxRetryTimeoutMillis) {
-            if (maxRetryTimeoutMillis <= 0) {
-                throw new IllegalArgumentException("maxRetryTimeoutMillis must be greater than 0");
-            }
-            this.maxRetryTimeout = maxRetryTimeoutMillis;
-            return this;
+        void onSuccess(Response response) {
+            responseListener.onSuccess(response);
         }
 
-        /**
-         * Sets the default request headers, to be used sent with every request unless overridden on a per request basis
-         */
-        public Builder setDefaultHeaders(Header[] defaultHeaders) {
-            Objects.requireNonNull(defaultHeaders, "defaultHeaders must not be null");
-            for (Header defaultHeader : defaultHeaders) {
-                Objects.requireNonNull(defaultHeader, "default header must not be null");
-            }
-            this.defaultHeaders = defaultHeaders;
-            return this;
+        void onDefinitiveFailure(Exception exception) {
+            trackFailure(exception);
+            responseListener.onFailure(this.exception);
         }
 
-        /**
-         * Sets the {@link FailureListener} to be notified for each request failure
-         */
-        public Builder setFailureListener(FailureListener failureListener) {
-            Objects.requireNonNull(failureListener, "failureListener must not be null");
-            this.failureListener = failureListener;
-            return this;
+        void trackFailure(Exception exception) {
+            this.exception = addSuppressedException(this.exception, exception);
         }
+    }
 
-        /**
-         * Sets the {@link HttpClientConfigCallback} to be used to customize http client configuration
-         */
-        public Builder setHttpClientConfigCallback(HttpClientConfigCallback httpClientConfigCallback) {
-            Objects.requireNonNull(httpClientConfigCallback, "httpClientConfigCallback must not be null");
-            this.httpClientConfigCallback = httpClientConfigCallback;
-            return this;
-        }
+    private static class SyncResponseListener implements ResponseListener {
+        final CountDownLatch latch = new CountDownLatch(1);
+        volatile Response response;
```

Comment:
do we have a `SetOnce` construct here? I think we should add it if not? or use AtomicReference and cmpSet and fail if it's already set?